File: android-app/app/src/main/java/com/mobileclaw/app/memory/InMemoryMemoryRepository.kt
package com.mobileclaw.app.memory

/**
 * In-memory, non-persistent MemoryRepository.
 *
 * Used by the eval harness (EvalActivity) so each eval task runs with a fresh,
 * isolated memory. The planner → evaluator → replan loop inside a single task
 * still sees and writes memory (episodes, repairs, device facts) the same way
 * the production repository would; the state is just discarded when the task ends.
 *
 * Holds everything in process memory — do not use for production user chat.
 */
class InMemoryMemoryRepository : MemoryRepository {
  private val lock = Any()
  private val episodes = mutableListOf<Episode>()
  private val repairs = mutableListOf<RepairRecord>()
  private val deviceFacts = mutableListOf<DeviceFact>()

  override suspend fun saveEpisode(episode: Episode) {
    synchronized(lock) { episodes.add(episode) }
  }

  override suspend fun saveRepair(repair: RepairRecord) {
    synchronized(lock) { repairs.add(repair) }
  }

  override suspend fun saveDeviceFact(key: String, value: String, sourceEpisodeId: String?) {
    synchronized(lock) {
      deviceFacts.removeAll { it.factKey == key }
      deviceFacts.add(
        DeviceFact(
          id = java.util.UUID.randomUUID().toString(),
          factKey = key,
          factValue = value,
          sourceEpisodeId = sourceEpisodeId,
        )
      )
    }
  }

  override suspend fun recallForPlanning(userMessage: String, tokenBudget: Int): String =
    synchronized(lock) {
      if (episodes.isEmpty() && repairs.isEmpty() && deviceFacts.isEmpty()) return ""
      val sb = StringBuilder()
      if (deviceFacts.isNotEmpty()) {
        sb.append("Device facts:\n")
        for (f in deviceFacts.takeLast(10)) sb.append("- ${f.factKey}: ${f.factValue}\n")
      }
      if (episodes.isNotEmpty()) {
        sb.append("Recent episodes:\n")
        for (e in episodes.takeLast(5)) {
          sb.append("- ${e.goal} → ${e.outcome} (skills=${e.skillsUsed.joinToString(",")})\n")
        }
      }
      if (repairs.isNotEmpty()) {
        sb.append("Recent repairs:\n")
        for (r in repairs.takeLast(3)) {
          sb.append("- ${r.skillName}: ${r.errorSummary} → ${r.fixDescription}\n")
        }
      }
      sb.toString()
    }

  override suspend fun recallRepairs(skillName: String, error: String, limit: Int): List<RepairRecord> =
    synchronized(lock) {
      val key = error.lowercase()
      repairs
        .asReversed()
        .filter { it.skillName == skillName && (key.isEmpty() || it.errorSummary.lowercase().contains(key)) }
        .take(limit)
    }

  override suspend fun getDeviceFacts(): List<DeviceFact> = synchronized(lock) { deviceFacts.toList() }

  override suspend fun evictIfNeeded(maxSizeBytes: Long) { /* in-memory; nothing to evict */ }

  override suspend fun clearAll() {
    synchronized(lock) {
      episodes.clear()
      repairs.clear()
      deviceFacts.clear()
    }
  }
}

File: ios-app-wip/MobileClaw/Memory/InMemoryMemoryRepository.swift
import Foundation

// 1:1 port of android-app/.../memory/InMemoryMemoryRepository.kt — in-memory
// store used by the eval harness and unit tests so a run doesn't persist
// learning across runs.

public actor InMemoryMemoryRepository: MemoryRepository {
    private var episodes: [Episode] = []
    private var repairs: [RepairRecord] = []
    private var facts: [String: DeviceFact] = [:]

    public init() {}

    public func save(episode: Episode) async { episodes.append(episode) }
    public func save(repair: RepairRecord) async { repairs.append(repair) }

    public func saveDeviceFact(key: String, value: String, sourceEpisodeId: String?) async {
        facts[key] = DeviceFact(factKey: key, factValue: value, sourceEpisodeId: sourceEpisodeId)
    }

    public func recallForPlanning(userMessage: String, tokenBudget: Int) async -> String {
        let needle = userMessage.lowercased()
        let recent = episodes.suffix(20)
        let related = recent.filter { ep in
            ep.userMessage.lowercased().contains(needle) || ep.goal.lowercased().contains(needle)
        }.prefix(3)
        if related.isEmpty { return "" }
        return related.map { "Past: \($0.userMessage) → \($0.outcome) (\($0.skillsUsed.joined(separator: ", ")))" }
            .joined(separator: "\n")
    }

    public func recallRepairs(skillName: String, error: String, limit: Int) async -> [RepairRecord] {
        let needle = error.lowercased()
        return repairs.reversed().filter {
            $0.skillName == skillName && $0.errorSummary.lowercased().contains(needle)
        }.prefix(limit).map { $0 }
    }

    public func getDeviceFacts() async -> [DeviceFact] { Array(facts.values) }

    public func evictIfNeeded(maxSizeBytes: Int64) async {
        // Approximate sizing: cap counts. Faithful eviction lands when the
        // SwiftData impl needs it for real disk pressure.
        if episodes.count > 1000 { episodes.removeFirst(episodes.count - 1000) }
        if repairs.count > 1000 { repairs.removeFirst(repairs.count - 1000) }
    }

    public func clearAll() async {
        episodes.removeAll(); repairs.removeAll(); facts.removeAll()
    }
}
